testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2250 @@
+
+static void con_init_dgram(struct con *conp)
+{
+	int server = SAFE_SOCKET(PF_INET, SOCK_DGRAM, 0);
+	struct sockaddr_in sin;
+	sockaddr_in_init(&sin, AF_INET, TEST_PORT, TEST_IP);
+	SAFE_BIND(server, (struct sockaddr *) &sin, sizeof(sin));
+	int client = SAFE_SOCKET(PF_INET, SOCK_DGRAM, 0);
+	con_init(conp, client, server, SOCK_DGRAM, &sin);
+}
+
+static void con_deinit(struct con *conp)
+{
+	close(conp->con_client);
+	close(conp->con_server);
+}
+
+static void con_add_send_recv_calls(struct con *conp, struct sendmsg_call *smcp,
+				    struct recvmsg_call *rmcp,
+				    bool sender_sleeps)
+{
+	if (conp->con_call_n > 0) {
+		con_add_send_recv_calls_vec(conp, smcp, rmcp, sender_sleeps);
+		return;
+	}
+
+	thread_call(&smcp->smc_thread_call, (void *(*)(void *))
+		    (sender_sleeps ? sendmsg_call_after_sleep : sendmsg_call));
+	struct timespec pretv, postv;
+	if (sender_sleeps)
+		tst_fzsync_time(&pretv);
+	ssize_t val = recvmsg(conp->con_server, rmcp->rmc_msg, rmcp->rmc_flags);
+	if (val < 0)
+		rmcp->rmc_errno = errno;
+
+	if (sender_sleeps) {
+		/*
+		 * If the sender is supposed to sleep prior to sending then
+		 * the receiver is supposed to sleep too because its receive
+		 * should block for the sender, testing that sender blocked
+		 * for at least half as long as the sender sleeps is good
+		 * enough to ensure that scheduling noise doesn't affect test.
+		 */
+		tst_fzsync_time(&postv);
+		unsigned long long pre = pretv.tv_sec * 1000 * 1000 * 1000;
+		pre += pretv.tv_nsec;
+		unsigned long long pos = postv.tv_sec * 1000 * 1000 * 1000;
+		pos += postv.tv_nsec;
+		test_assert((pos - pre) > (SENDMSG_SLEEP_MS * 1000) / 2);
+	}
+
+	rmcp->rmc_value = val;
+	thread_join(&smcp->smc_thread_call);
+}
+
+/*
+ *	Miscellaneous supporting code is in this section.
+ */
+
+static int errno_value_is_error(int e)
+{
+	return e > 0;  /* negative values are not errors, see non_error_errno */
+}
+
+static void sleep_ms(long ms)
+{
+	test_assert(ms < 1000);
+	struct timespec ts;
+	ts.tv_sec = 0;
+	ts.tv_nsec = ms * 1000 * 1000;
+	nanosleep(&ts, NULL);
+}
+
+static int test_number_get(void)
+{
+	static int test_number_gen;
+
+	/*
+	 * Nested subtests share the same test number, these are multi
+	 * message tests, i.e. recvmmsg() and sendmmsg(), where each message
+	 * is treated as a subtest
+	 */
+
+	if (subtest_number() > 1)
+		return test_number_gen;
+	return ++test_number_gen;
+}
+
+/*
+ * subtests numbers
+ */
+
+static int subtest_level = 1;
+
+static int subtest_number(void)
+{
+	return subtest_level;
+}
+
+static void subtest_nest(void)
+{
+	++subtest_level;
+}
+
+static void subtest_unnest(void)
+{
+	--subtest_level;
+}
+
+static bool subtest_show(void)
+{
+	/*
+	 *	Some tests involve too many subtests, this limits the output.
+	 */
+	return subtest_level <= 20;
+}
+
+/*
+ * compare that memory is equal, makes the output of various ensure() nicer
+ */
+
+static bool mem_is_equal(void *a, void *b, size_t size)
+{
+	return !memcmp(a, b, size);
+}
```

Comment:
Can we avoid wrappers around standard C functions like this? The memcmp() _is_ the function to compare memory. Wrappers like this not only inflate the source code but also makes it harder to read.